### ISSUE TITLE
"undefined method `+' for false:FalseClass" when using package resources on RHEL 8

### ISSUE THREAD

**Issue by Poohblah** [chef/chef#7988]
## Description
`package`, `yum_package`, and `dnf_package` all fail to install packages on RHEL 8.

## Chef Version
15.0.59

Also tested 14.7.17 and 14.2.0 and saw the same behavior on all three versions.

## Platform Version
RHEL 8.0 beta 1

## Replication Case

Create a recipe that installs a package, for instance:

```
package('yum-plugin-versionlock')
```

I tried `package`, `dnf_package`, and `yum_package`.  All three throw errors.

## Client Output

When using `package` or `dnf_package`:

       Starting Chef Client, version 15.0.59
       Using policy 'package_bugs' at revision 'ed2accc3ccada472522f9c4a4551b015c58e704eb403c08d6a6f37f7b1e8518c'
       resolving cookbooks for run list: ["package_bugs::default@0.1.0 (a45a459)"]
       Synchronizing Cookbooks:
         - package_bugs (0.1.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: package_bugs::default
         * dnf_package[yum-plugin-versionlock] action install
           
           ================================================================================
           Error executing action `install` on resource 'dnf_package[yum-plugin-versionlock]'
           ================================================================================
           
           NoMethodError
           -------------
           undefined method `+' for false:FalseClass
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/package_bugs/recipes/default.rb
           
             1: package('yum-plugin-versionlock')
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/package_bugs/recipes/default.rb:1:in `from_file'
           
           dnf_package("yum-plugin-versionlock") do
             package_name "yum-plugin-versionlock"
             action [:install]
             default_guard_interpreter :default
             declared_type :package
             cookbook_name "package_bugs"
             recipe_name "default"
           end
           
           System Info:
           ------------
           chef_version=15.0.59
           platform=redhat
           platform_version=8.0
           ruby=ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]
           program_name=/opt/chef/bin/chef-client
           executable=/opt/chef/bin/chef-client
           
       
       Running handlers:
       [2018-11-27T14:43:28-07:00] ERROR: Running exception handlers
       Running handlers complete
       [2018-11-27T14:43:28-07:00] ERROR: Exception handlers complete
       Chef Client failed. 0 resources updated in 01 seconds
       [2018-11-27T14:43:28-07:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2018-11-27T14:43:28-07:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2018-11-27T14:43:28-07:00] FATAL: NoMethodError: dnf_package[yum-plugin-versionlock] (package_bugs::default line 1) had an error: NoMethodError: undefined method `+' for false:FalseClass

`yum_package` throws a very similar error:

       Starting Chef Client, version 15.0.59
       Using policy 'package_bugs' at revision 'a69b6542f0399110a06cabd0dbca33bc850e8910902e56be2b83c3eb46b3c62e'
       resolving cookbooks for run list: ["package_bugs::default@0.1.0 (4e99a6c)"]
       Synchronizing Cookbooks:
         - package_bugs (0.1.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: package_bugs::default
         * yum_package[yum-plugin-versionlock] action install
           
           ================================================================================
           Error executing action `install` on resource 'yum_package[yum-plugin-versionlock]'
           ================================================================================
           
           NoMethodError
           -------------
           undefined method `+' for false:FalseClass
           
           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/package_bugs/recipes/default.rb
           
             1: yum_package('yum-plugin-versionlock')
           
           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/package_bugs/recipes/default.rb:1:in `from_file'
           
           yum_package("yum-plugin-versionlock") do
             package_name "yum-plugin-versionlock"
             action [:install]
             default_guard_interpreter :default
             declared_type :yum_package
             cookbook_name "package_bugs"
             recipe_name "default"
           end
           
           System Info:
           ------------
           chef_version=15.0.59
           platform=redhat
           platform_version=8.0
           ruby=ruby 2.5.3p105 (2018-10-18 revision 65156) [x86_64-linux]
           program_name=/opt/chef/bin/chef-client
           executable=/opt/chef/bin/chef-client
           
       
       Running handlers:
       [2018-11-27T14:44:33-07:00] ERROR: Running exception handlers
       Running handlers complete
       [2018-11-27T14:44:33-07:00] ERROR: Exception handlers complete
       Chef Client failed. 0 resources updated in 01 seconds
       [2018-11-27T14:44:33-07:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2018-11-27T14:44:33-07:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2018-11-27T14:44:33-07:00] FATAL: NoMethodError: yum_package[yum-plugin-versionlock] (package_bugs::default line 1) had an error: NoMethodError: undefined method `+' for false:FalseClass

## Stacktrace

```
Generated at 2018-11-27 14:50:08 -0700
NoMethodError: dnf_package[yum-plugin-versionlock] (package_bugs::default line 1) had an error: NoMethodError: undefined method `+' for false:FalseClass
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:42:in `dnf_command'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:47:in `start'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:61:in `check'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:147:in `block in with_helper'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:93:in `block in timeout'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `block in catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:108:in `timeout'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:146:in `with_helper'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:75:in `query'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:157:in `installed_version'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:92:in `block in get_current_versions'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `map'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `get_current_versions'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:69:in `load_current_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider.rb:165:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource.rb:578:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:70:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:97:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:720:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:715:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:715:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:754:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:286:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:279:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:261:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application/client.rb:440:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:66:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/bin/chef-client:25:in `<top (required)>'
/opt/chef/bin/chef-client:74:in `load'
/opt/chef/bin/chef-client:74:in `<main>'

>>>> Caused by NoMethodError: undefined method `+' for false:FalseClass
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:42:in `dnf_command'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:47:in `start'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:61:in `check'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:147:in `block in with_helper'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:93:in `block in timeout'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `block in catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:33:in `catch'
/opt/chef/embedded/lib/ruby/2.5.0/timeout.rb:108:in `timeout'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:146:in `with_helper'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf/python_helper.rb:75:in `query'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:157:in `installed_version'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:92:in `block in get_current_versions'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `map'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:91:in `get_current_versions'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider/package/dnf.rb:69:in `load_current_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/provider.rb:165:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource.rb:578:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:70:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `block (2 levels) in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `each'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:98:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/resource_list.rb:94:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:114:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:103:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/resource_collection/resource_list.rb:92:in `execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/runner.rb:97:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:720:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:715:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:715:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:754:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/client.rb:286:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:303:in `run_with_graceful_exit_option'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:279:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/local_mode.rb:44:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:261:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application/client.rb:440:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/lib/chef/application.rb:66:in `run'
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/chef-15.0.59/bin/chef-client:25:in `<top (required)>'
/opt/chef/bin/chef-client:74:in `load'
/opt/chef/bin/chef-client:74:in `<main>'
```

**Comment by pixdrift** [comment]
@Poohblah, I just hit this issue also.

I believe the problem is that default RHEL 8 doesn't ship a userspace python install (installed in the path), so the python_helper falls over here when it attempts to find a viable python binary to import the dnf library here:

https://github.com/chef/chef/blob/master/lib/chef/provider/package/dnf/python_helper.rb#L39-L43
```
          def dnf_command
            @dnf_command ||= which("python", "python3", "python2", "python2.7") do |f|
              shell_out("#{f} -c 'import dnf'").exitstatus == 0
            end + " #{DNF_HELPER}"
          end
```

I resolved the above issue with Chef by installing a python binary on the RHEL 8 system before the chef-client run (which shows it's an issue with python binary location), eg.
```
yum install python3
```

There is an additional option. RHEL 8 ships a python binary that it uses for system tools (including DNF) hidden/abstracted away from the end user in the `platform-python` RPM. This lives in the following location and is not added to the path:
```
[root@rhel8b1 ~]# /usr/libexec/platform-python --version
Python 3.6.6
[root@rhel8b1 ~]# rpm -qf /usr/libexec/platform-python
platform-python-3.6.6-19.el8.x86_64
```

I think Chef should probably be using this python binary on RHEL 8 to load the helper.

**Comment by Poohblah** [comment]
@pixdrift you are absolutely right.  I also looked at the source code for the `dnf_command` and `yum_command` methods but failed to draw the full conclusion.  Installing userspace Python 2 or Python 3 resolves the error.  I agree that it would be best for Chef to use platform-python rather than userspace python.

I took a look at the code for the `which` method as well.  This method assumes that any executable will be in one of the directories `[ "/bin", "/usr/bin", "/sbin", "/usr/sbin" ]`.  So this method may need to be patched too.
